src/main/scala/rise/eqsat/ExtendedPattern.scala

```diff
@@ -0,0 +1,753 @@
+        dts.zip(pdts).forall { case (dt, pdt) => dataTypeIsMatch(egraph, pdt, dt) } &&
+        ts.zip(pts).forall { case (t, pt) => typeIsMatch(egraph, pt, t) }
+    }
+  }
+  def natIsMatch(egraph: EGraph, p: NatPattern, t: NatId): Boolean = {
+    p match {
+      case NatPatternNode(pn) =>
+        val n = egraph(t)
+        (pn.map(_ => ()) == n.map(_ => ())) && {
+          val ns = Vec.empty[NatId]
+          n.map(n => ns += n)
+          val pns = Vec.empty[NatPattern]
+          pn.map(pn => pns += pn)
+          ns.zip(pns).forall { case (n, pn) => natIsMatch(egraph, pn, n) }
+        }
+      case NatPatternVar(index) => ??? // TODO
+      case NatPatternAny => true
+    }
+  }
+  def addressIsMatch(p: AddressPattern, t: Address): Boolean = {
+    p match {
+      case AddressPatternNode(pn) => pn == t
+      case AddressPatternVar(index) => ??? // TODO
+      case AddressPatternAny => true
+    }
+  }
+
+  def traverse[A, B](iterator: Iterator[A])(f: A => Option[B]): Option[Vec[B]] = {
+    val acc = Vec.empty[B]
+    while (true) {
+      iterator.nextOption() match {
+        case Some(a) => f(a) match {
+          case Some(b) => acc += b
+          case None => return None
+        }
+
+        case None => return Some(acc)
+      }
+    }
+
+    throw new Exception("unreachable")
+  }
+}
+
+sealed trait ExtendedPatternMatch
+object ExtendedPatternMatch {
+  type MNode[T] = Node[T, NatId, DataTypeId, Address]
+
+  /** The entire e-class is a match */
+  case class EClass(id: EClassId) extends ExtendedPatternMatch
+
+  /** Each node child has multiple matches, creating a cartesian product of matches.
+    *
+    *   P(f({a, b}, {c, d})) = {f(a, c), f(a, d), f(b, c), f(b, d)}
+    */
+  case class Product(node: MNode[Vec[ExtendedPatternMatch]], id: EClassId) extends ExtendedPatternMatch
+
+  /** Traversing each node child gives multiple matches.
+    * For each possible traversal match from `traverse`,
+    * the non-traversed children from `node` should remain the original e-classes.
+    *
+    *   TP(f(x, y), {{a, b}, {c, d}}) = {f(x, c), f(x, d), f(a, y), f(b, y)}
+    */
+  case class TraversalProduct(node: ENode, traverse: Vec[Vec[ExtendedPatternMatch]], id: EClassId) extends ExtendedPatternMatch
+}
+
+/** An extended pattern */
+sealed trait ExtendedPattern {
+  import ExtendedPattern.{typeIsMatch, dataTypeIsMatch, natIsMatch, addressIsMatch, traverse}
+
+  def searchEClass(egraph: EGraph, id: EClassId): Vec[ExtendedPatternMatch] = {
+    assert(egraph.clean)
+    searchEClass(egraph, egraph.get(id), Set(), HashMap.empty/*, None*/)
+  }
+
+  // TODO: investigate similarities with e-matching
+  private def searchEClass(egraph: EGraph,
+                           eclass: EClass,
+                           visited: Set[EClassId],
+                           memo: HashMap[(EClassId, ExtendedPattern), Vec[ExtendedPatternMatch]],
+                           /*parent: Option[ENode]*/): Vec[ExtendedPatternMatch] = {
```

Comment:
This search method is technically not used anymore .. should we keep the code around or delete it?